src/coordinator/tsdb/remote/codecs.go

```diff
@@ -45,46 +45,79 @@ func EncodeFetchResult(sResult *storage.FetchResult) *rpc.FetchResult {
 	series := make([]*rpc.Series, len(sResult.SeriesList))
 	for i, result := range sResult.SeriesList {
 		vLen := result.Len()
-		vals := make([]float32, vLen)
+		datapoints := make([]*rpc.Datapoint, vLen)
+		_, fixedRes := result.Values().(ts.FixedResolutionMutableValues)
 		for j := 0; j < vLen; j++ {
-			vals[j] = float32(result.ValueAt(j))
+			dp := result.Values().DatapointAt(j)
+			datapoints[j] = &rpc.Datapoint{
+				Timestamp: fromTime(dp.Timestamp),
+				Value:     dp.Value,
+			}
 		}
+
 		series[i] = &rpc.Series{
-			Name:          result.Name(),
-			Values:        vals,
-			StartTime:     fromTime(result.StartTime()),
-			Tags:          result.Tags,
-			Specification: result.Specification,
-			MillisPerStep: int32(result.MillisPerStep()),
+			Name: result.Name(),
+			Values: &rpc.Datapoints{
+				Datapoints:      datapoints,
+				FixedResolution: fixedRes,
+			},
+			StartTime: fromTime(result.StartTime()),
+			Tags:      result.Tags,
 		}
 	}
 	return &rpc.FetchResult{Series: series}
 }
 
 // DecodeFetchResult decodes fetch results from a GRPC-compatible type.
-func DecodeFetchResult(ctx context.Context, rpcSeries []*rpc.Series) []*ts.Series {
+func DecodeFetchResult(_ context.Context, rpcSeries []*rpc.Series) []*ts.Series {
 	tsSeries := make([]*ts.Series, len(rpcSeries))
 	for i, series := range rpcSeries {
-		tsSeries[i] = decodeTs(ctx, series)
+		tsSeries[i] = decodeTs(series)
 	}
 	return tsSeries
 }
 
-func decodeTs(ctx context.Context, r *rpc.Series) *ts.Series {
-	millis, rValues := int(r.GetMillisPerStep()), r.GetValues()
-	values := ts.NewValues(ctx, millis, len(rValues))
+func decodeTs(r *rpc.Series) *ts.Series {
+	fixedRes := r.Values.FixedResolution
+	var values ts.Values
+	if fixedRes {
+		values = decodeFixedResTs(r)
 
-	for i, v := range rValues {
-		values.SetValueAt(i, float64(v))
+	} else {
+		values = decodeRawTs(r)
 	}
 
 	start, tags := toTime(r.GetStartTime()), models.Tags(r.GetTags())
-
-	series := ts.NewSeries(ctx, r.GetName(), start, values, tags)
-	series.Specification = r.GetSpecification()
+	series := ts.NewSeries(r.GetName(), start, values, tags)
 	return series
 }
 
+func decodeFixedResTs(
+	r *rpc.Series) ts.FixedResolutionMutableValues {
+	startTime := time.Now()
+	if len(r.Values.Datapoints) > 0 {
+		startTime = toTime(r.Values.Datapoints[0].Timestamp)
+	}
+
+	values := ts.NewFixedStepValues(millisPerStep(r.Values), len(r.Values.Datapoints), 0, startTime)
```

Comment:
nit: change the constructor to take the `Datapoints` rather than constructing it manually here?

Reply:
makes some tests easier to have it separate, also the assumption is that ts.NewFixedStepValues() will give back pooled values on which we would later set the datapoints.